src/utils.ts

```diff
@@ -106,3 +108,17 @@ export const getZoneLabel = (zone?: any) => {
 
   return label
 }
+
+export const migrateConfiguration = (configuration: any, version: number) => {
+  let updatedConfiguration = { ...configuration }
```

Comment:
Why not just `let updatedConfiguration = configuration`?

Reply:
Because `let updatedConfiguration = configuration` doesn't makes a copy of `configuration` (it just makes a pointer), and then I might as well not make a new variable at all. I'd actually prefer not to make a new variable, but then I'd have to reassign the function's argument `configuration`, which causes the lint error: `ESLint: Assignment to function parameter 'configuration'.(no-param-reassign)`. 

As it happens, this function won't mutate configuration whether I use `let updatedConfiguration = configuration` or `let updatedConfiguration = { ...configuration }`. But I still don't like `let updatedConfiguration = configuration` because from that line alone it's heavily implied I'm going to make changes to `updatedConfiguration` later in the function, but it's not clear if those changes are going to mutate `configuration`.

For example, these two functions return the same thing (and both mutate their inputs):
```
const func1 = (obj) => {
  obj.prop = 'hello' // lint error
  return obj
}
```

```
const func2 = (obj) => {
  const newObj = obj
  newObj.prop = 'hello' // no lint error
  return obj
}
```

I think if the first line in `func1` has a lint error, then the first line in `func2` should throw a different lint error - it should tell you to write `newObj = { ...obj }` (or equivalent) instead. What do you think?

Reply:
> As it happens, this function won't mutate configuration

Exactly.

> But I still don't like let updatedConfiguration = configuration because from that line alone it's heavily implied I'm going to make changes to updatedConfiguration later in the function, but it's not clear if those changes are going to mutate configuration.

I disagree. `let` only implies you'll reassign `updatedConfiguration`. In fact it implies you _won't_ mutate the object itself. If you were only mutating (not reassigning) the object, you could use `const`.

This is perfectly valid:

```
const obj = otherObj
obj.thing = 5
```

`let updatedConfiguration = { ... configuration }` implies you plan to both reassign `updatedConfiguration` _and_ mutate the object, which isn't what's happening.

Reply:
I think these are a bit at odds with each other:

> `let` only implies you'll reassign `updatedConfiguration`. In fact it implies you won't mutate the object itself.

> `let updatedConfiguration = { ... configuration }` implies you plan to both reassign `updatedConfiguration` and mutate the object

...but I can handle the nuance: `const` implies _only_ mutation without reassign (duh, it's strictly enforced), and `let` implies _only_ reassign without mutation... unless it's `let thing =  {...otherThing}`, which implies both.

To your other point, I still think this is problematic when `otherObj` is the function parameter:
```
const obj = otherObj
obj.thing = 5
```
Because `obj.thing = 5` also makes `otherObj.thing = 5`, which you're not supposed to do.

Reply:
> I still think this is problematic when otherObj is the function parameter:

Depends on the function. But in many cases, yes, you want to avoid mutating the argument. But that's an argument against mutating the argument, not an argument for coping an object you're not planning to mutate.

Reply:
> But that's an argument against mutating the argument, not an argument for coping an object you're not planning to mutate.

Can you give an example where I'd copy an object I didn't want to mutate? (and won't reassign because I used `const`) Wouldn't I just use the original object instead?

Reply:
> Can you give an example where I'd copy an object I didn't want to mutate? (and won't reassign because I used const)

You may copy an object that you expect to be mutated elsewhere. But mostly, you wouldn't copy an object you weren't planning to mutate, which is why that line struck me as odd.

Reply:
I agree, you wouldn't copy an object you weren't planning to mutate.

Reply:
Got it. I thought I was arguing to avoid mutating the argument, but I was actually arguing to copy stuff willy nilly.